lib/Dialect/Torch/Transforms/GlobalizeObjectGraph.cpp

```diff
@@ -215,15 +241,26 @@ class ObjectGraphInfo {
     BlockAndValueMapping mapping;
     for (Operation *op : worklist) {
       builder.clone(*op, mapping);
-      for (Value result : op->getResults()) {
+      for (auto result_it : llvm::enumerate(op->getResults())) {
+        Value result = result_it.value();
         if (!hasMeaningfulObjectIdentity(result.getType()))
           continue;
-        if (usedSlots.find(slot) == usedSlots.end())
+        if (slotUses.find(slot) == slotUses.end())
           continue;
         if (!objectsWithIdentityAlreadyCopiedIntoInitializers.insert(result)
                  .second) {
-          return op->emitError() << "potentially-aliased value used to "
-                                    "initialize multiple slots";
+          // Check whether the value's uses ever create aliases or mutate.
+          if (failed(checkUsesForValueSemantics(result)))
+            return op->emitError()
+                   << "unsafe initializer used to initialize multiple slots";
+
+          // Create a deep clone of the `result` value before inserting it.
+          Operation *clonedOp = op->clone();
```

Comment:
cloning seems like it could be a problem if it is a large weight buffer -- in practice, how big are the things you are cloning? And do some of the clones get deleted?

Reply:
Sadly, for the model that I'm compiling, the buffers are large (some tens of millions of values), but that would, of course, change with the model.  I haven't studied the uniqueness requirement, so pardon by ignorance, but will it be a significant undertaking to remove the non-aliasing requirement for initializers?

Reply:
Is there a way for you to construct a reduced test case that I could poke at?

Reply:
Sure, here's one starting from Python: https://gist.github.com/ashay/5f26017affc81fe0eaba81b2935bc950.  The comment on that gist shows the error message.  Thanks for offering to look into this!

Reply:
Thanks! I have a solution, but it makes torch.global_slot initializers quite a bit more complex for passes to handle. Basically we allow the initializer for one slot to reference another slot. This makes InlineGlobalSlots quite a bit more complex.

So, I'm trying to understand the need for this feature better to weigh it against the increase in complexity of the system. Can you explain a bit more how this occurs "in the wild"? Is this just a pathological example, or is there some common pattern that will end up triggering this from many users?

Reply:
Sorry, this took a while to comb through the model source code.  It seems as though the key culprit is a reference to `self.parameters()` in the top-level module.  I'm having a hard time reproducing the exact problem in a smaller example, but I'll let you know once I succeed.

Reply:
The problem is that `self.parameters()` refers to not just the parameters of the calling `torch.nn.Module` but also the parameters of the member fields.  [Here](https://gist.github.com/ashay/15ecab16886356424e27b60649328da1) is an example, which creates an embedding layer, but the top-level module reaches into the embedding layer to use its weight parameter.

Reply:
Okay, that seems pretty reasonable. I will work on a fix for this.

Reply:
Started work on this in https://github.com/llvm/torch-mlir/pull/1047 PTAL